### PR TITLE
Filter by status and stage

### PR DESCRIPTION
**Pull request by bytheway875**
Screen Shot 2020-11-16 at 2 51 08 PM
Screen Shot 2020-11-16 at 2 56 08 PM
Screen Shot 2020-11-16 at 2 56 25 PM

### CODE REVIEW

File: app/lib/tax_return_status.rb
Comment (bytheway875) [inline app/lib/tax_return_status.rb:1]: @alaroia-cfa @sartan There are a couple of reasons that I resurrected the TaxReturnStatus class, albeit in a more simplified form.

First, I'm always trying to keep the single responsibility principle + Sandi Metz' "rules" in mind (https://thoughtbot.com/blog/sandi-metz-rules-for-developers) when writing Ruby classes, and one of the big ones is to keep classes less than 100 lines. The TaxReturn model was pushing 100 lines, and most of it had to do with methods dealing only with setting up statuses and stages. 

Second, because of the way I'm trying to calculate the value of a const outside of the context of a method using a method, the method needed to be defined BEFORE it was used in the class, and it was really strange to do that in TaxReturn class itself. 

Third, since we're using the stages in backend code to create scopes and facilitate filtering, it made sense to me to have the calculation of the stages outside of the helper which are more for presentational purposes. The determine_statuses_by_stage method is really similar to what was in the helper, but now it returns the raw stage name, which then gets translated in the helper.

File: app/models/client.rb
Comment (bytheway875) [inline app/models/client.rb:45]: In order to use "distinct", we need to include the intake properties in delegated_intake_attributes in our select. We _could_ limit it to just the column we're passing in, but I was worried it might trigger a Hakiri violation because we'd be creating a query from user-provided data.

File: app/views/shared/_column_sort_link.erb
Comment (bytheway875) [inline app/views/shared/_column_sort_link.erb:3]: Allows us to maintain the existing filters when we sort by a column.

File: app/views/shared/_client_filters.erb
Comment (bytheway875) [inline app/views/shared/_client_filters.erb:17]: Passing in the sort order and column as hidden fields allows us to maintain the current order

File: app/controllers/concerns/client_sortable.rb
Comment (bytheway875) [inline app/controllers/concerns/client_sortable.rb:27]: I'm hoping this creates a pattern we can use to add scopes to the query as we add new filters in future tickets.

File: app/lib/tax_return_status.rb
Comment (bytheway875) [reply app/lib/tax_return_status.rb:1]: I'm also willing to move this back into TaxReturn -- I actually haven't moved the tests because I'm waiting for feedback before making a final decision.

**Comment by sartan** [comment]
We're seeing the dropdown cleared after applying a filter.

image

**Review by alaroia-cfa — APPROVED** [review]
This does the job and is great! We left a comment with one bug. Happy to have you make minor changes; feel free to request review or just merge with minor changes if the tests still pass.

**Comment by alaroia-cfa** [comment]
to do list if Alex & Asheesh grab this

- use explicit method call instead of `.send`
- fix `selected` issue & write feature test for it
- use `params[:status]` instead of `@filters` in `link_to`
- see if we can stop `render_views: false`
- use `TaxReturnStatus` constants directly instead of `TaxReturn` method, maybe requiring some tests to move
- Use `[] + something.map` for delegated order

**Comment by sartan** [comment]
Needs further investigation, but possibly a bug relating to assignment/tax year.

tax_year_filter_bug

**Comment by alaroia-cfa** [comment]
It's green, and I'm gonna merge it!

**Pull request merged**
